### ISSUE TITLE
[Bug] Mutating State directly

### ISSUE THREAD

**Issue by bchuntley** [StephenChou1017/react-big-scheduler#15]
As far as I've traced it, state is modified directly and the object is never re-declared or over-written. This had led to feature developments and new implementations difficult, as I can't trust that the original objects I supply to `schedulerData` will not be modified.

**Comment by StephenChou1017** [comment]
@bchuntley
Yeah, that's a problem, I didn't consider it when designing. But the scheduler component is simple:), the Scheduler JSX is view, the SchedulerData is view model, it has a few core properties: time window, view type, config, resources, events, and it regenerates renderData the view need every time after resources or events change, usually we concern the events data, we should ensure the events are not unexpectedly modified, so in  my practice I'll commit the modified event to server after editing it, re-request the events data and refresh the view.

Can rendering the component from server side resolve the problem you mentioned, or any suggestions?

**Comment by bchuntley** [comment]
The fix I'm running with is actually managing the events outside of the scheduler and removing the scheduler data completely, and instead using memoiziation to calculate render data, that way you dont run into the issue of needing to regenerate _everything_ inside the scheduler for one change.

I think this fits a lot better in-line with React's docs because then its not inherintly tied to the blackbox'd SchedulerData object

**Comment by StephenChou1017** [comment]
Thanks for the reply, removing the scheduler data completely means many changes:-), I'll think about it.

**Comment by cmorrow-tg** [comment]
> The fix I'm running with is actually managing the events outside of the scheduler and removing the scheduler data completely, and instead using memoiziation to calculate render data, that way you dont run into the issue of needing to regenerate _everything_ inside the scheduler for one change.
> 
> I think this fits a lot better in-line with React's docs because then its not inherintly tied to the blackbox'd SchedulerData object

Would love to see your implementation of this if possible @bchuntley

**Comment by bchuntley** [comment]
@cmorrow-tg This was back at a company I'm no longer working at, but the code I wrote for it is publicly available [here](https://github.com/crossroads-education/element-web-scheduler). We ended up just rewriting it from scratch.
